### PR TITLE
Implement services (ADR-32)

### PR DESCRIPTION
**Pull request by caspervonb**
This implements services ([ADR-32](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-32.md)), this is based on a stripped down version of [this implementation](https://github.com/charbonats/nats-micro) by @charbonnierg.

This does not have a "client" or "monitor" type like nats.deno, we can come back and add a higher level interface for it later if it is needed. For the first pass it aims for functional parity with nats.go.

Also adds compatibility tests.

Closes https://github.com/nats-io/nats.py/issues/552

### CODE REVIEW

**Comment by wallyqs** [comment]
Think may have to update these: https://github.com/nats-io/nats.py/blob/main/setup.py#L14

**Comment by swelborn** [comment]
Thanks for making this -- appears to be working well! I do have a static checker issue with `micro.add_service`:

`"add_service" is not exported from module "nats.micro"`

Adding that to the list of exports in `micro.__init__.py` removes it.

```python
__all__ = [
    "api",
    "control_subject",
    "Verb",
    "Endpoint",
    "Group",
    "Handler",
    "Request",
    "Service",
    "add_service"
]
```

File: tests/test_micro.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+            svc = micro.add_service(
+                nc,
+                name="CoolAddService",
+                version="0.1.0",
+                description="Add things together",
+                metadata={"basic": "metadata"},
+            )
+
+            await svc.add_endpoint(
+                name="svc.add",
```
Comment (mgale) [inline tests/test_micro.py:36]: Running nats cli version 0.1.4:
```
nats --version
v0.1.4
```

Returns the following error on `nats micro list`:
```
nats: error: /endpoints/0/name: does not match pattern '^[a-zA-Z0-9_-]+$'
```

The endpoint `name` can not have a dot "."

**Comment by swelborn** [comment]
I noticed this as well @mgale. It can be resolved by adding a group to a
group.

Not sure if this is expected behavior or if it aligns with go micro API.
Would be cool if you could specify endpoints with dots and they would turn
into groups/subgroups.

Best,
Sam

**Comment by mgale** [comment]
I noticed an issue with nats-surveyor trying to parse the messages that go through the service, not sure where the failure is or if this works with the nats.go. 
```
[WARN] unsupported service observation received for id: /observations/obs.json, service name: myservice, kind: io.nats.unknown_message
```

From the suryeor code:
https://github.com/nats-io/nats-surveyor/blob/main/surveyor/observation.go#L358

File: nats/micro/request.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+# Copyright 2021-2022 The NATS Authors
```
Comment (wallyqs) [inline nats/micro/request.py:1]:
```suggestion
# Copyright 2021-2024 The NATS Authors
```
etc..

File: tests/test_micro_service.py [hunk 3]
```diff
@@ -0,0 +1,782 @@
...
+
+                assert ping.id
+                assert ping.name == expected_ping.name
+                assert ping.version == expected_ping.version
+                assert ping.metadata == expected_ping.metadata
+
+                await svc.stop()
+                assert svc.stopped
+
+    # async def test_err_handler_subject_match(self):
```
Comment (wallyqs) [inline tests/test_micro_service.py:220]: ok to leave commented?

File: nats/micro/service.py [hunk 4]
```diff
@@ -0,0 +1,848 @@
...
+        if self.subject:
+            if not SUBJECT_REGEX.match(self.subject):
+                raise ValueError(
+                    "Invalid subject. Subject must not contain spaces, and can only have '>' at the end."
+                )
+
+        if self.queue_group:
+            if not SUBJECT_REGEX.match(self.queue_group):
+                raise ValueError(
+                    "Invalid queue group. Queue group must not contain spaces, and can only have '>' at the end."
```
Comment (aricart) [inline nats/micro/service.py:90]: not sure what `>` is at the name of a queue group - this is copy/paste?

File: nats/micro/service.py [hunk 5]
```diff
@@ -0,0 +1,848 @@
...
+
+        if not NAME_REGEX.match(self.name):
+            raise ValueError(
+                "Invalid name. Name must contain only alphanumeric characters, underscores, and hyphens."
+            )
+
+        if self.subject:
+            if not SUBJECT_REGEX.match(self.subject):
+                raise ValueError(
+                    "Invalid subject. Subject must not contain spaces, and can only have '>' at the end."
```
Comment (aricart) [inline nats/micro/service.py:84]: is this error for the internal service subjects?

**Review by aricart — APPROVED** [review]
LGTM - only a question on some error messages

**Review by wallyqs — APPROVED** [review]
LGTM

**Pull request merged**